### ISSUE TITLE
Renaming a folder lose the focuses

### ISSUE THREAD

**Issue by nicolasmol** [quicken/filefocus#2]
- quit vscode
- rename a folder
- the focuses are losts

**Comment by quicken** [comment]
In the current implementation, the extension keeps track of file system locations.

A core feature is supporting multiple workspace folders. For example when you have two project folders added to your workspace and want to mix files from both projects in one focus group.

Because it is faster in terms of performance and more importantly coding effort focus group data is currently kept simple by storing the URI of the resource. That URI is in absolute terms and not relative to the project folder. 

Therefore, renaming a folder will mean the extension can no longer see the resource. That is what is happening when you rename a project. However, in the latest release, you should see a "warning" icon next to a missing resource.

As such the reported "bug" while no doubt annoying to you, is expected behaviour.


### Possible solution/feature
Now if I understand you correctly from this and your other reported issue, you are wanting a process based around focus groups being associated with a specific project (workspace folder) rather than an entire workspace. 

In this scenario, the focus group configuration would be saved inside of the project folder. As a consequence, you could then share focus groups and also have templating support. 

Of course, if the configuration is stored in the project that would mean that anyone using that project would have to have the same groups. However, this is also problematic when developers want to define their own personal groups.

Perhaps, the approach should be that a "default config file" can be defined inside the project that can be used as the basis for a project's default focus groups. Every developer can then continue to work as they want with the option to import the project default at any time. 

In addition, an extension preference could enable "global" storage that way groups could be optionally synced wherever you have the same account. However, this could be an issue where a workspace does not exist on a synced machine. eg. work vs home.

**Maybe, the way to have the cake and eat it is to have more "tree views".** 

That way in addition to the current tree view there could be a separate tree view that shows a project-specific set of "focus groups" with those focus groups being stored inside of the project folder.

In other words, a second tree view that works as you suggested.

Thoughts?



### Notes here for my own reference. 

Consider the following:

- if groups are stored in a project folder what happens when multiple projects are added to the workspace where focus group names overlap?
- build our own "memento" instead of leaning on the work done by vscode engineers.
- rework storage so that workspace location is stored separately from resource paths.
- how to deal with syncing between computers where paths will most likely be different. e.g. a work PC and home laptop.
- sacrifice the speed of simply looking up paths as paths will then need to be rebuilt to keep the tree view from lagging.

**Comment by nicolasmol** [comment]
Just take inspiration from Explorer-Exclude extension: https://github.com/sfccdevops/explorer-exclude-vscode-extension. It saves data in .vscode/settting.json and give someanswers to your questions.

https://github.com/sfccdevops/explorer-exclude-vscode-extension/blob/develop/extension/util.js

**Comment by quicken** [comment]
In the Beta, it is now possible to define a .filefocus.json file in the root of a workspace folder. 

This makes it possible to maintain and share a set of focus groups. Resources defined in a .filefocus.json file are relative to the file location. 

Therefore, renaming a root folder will maintain focus groups.

Note: 
Resources of Focus Groups that are stored within the workplace state or globally state (also new) will still not survive a rename as support for multiple workspace folders requires the workspace name to be a part of the resource. 

It may be possible to ignore the workspace folder name when there is "single workspace folder".

**Comment by quicken** [comment]
Fixed in 1.4.1 Beta 3. 
In a workspace with a single folder, the extension will try to automatically map resources based on a relative path. This allows configuring a Focus Group with files commonly in use by projects. 

In a multi workspace folder resources rely on the workspace name to be the same as the workspace from which the resource was added.

**Comment by nicolasmol** [comment]
Trying the 1.4.1 Beta 3.
How doest it work exactly ?
Is .filefocus.json auto-generated or manually ?
I tried both but it doesn't do anything
Here screen capture after restarting.
Note that I haves tried both `filefocus.useGlobalStorage `true/false
![image](https://user-images.githubusercontent.com/61077647/189194168-a0ee52ba-3f35-496c-a54c-f967b25bae3a.png)

**Comment by quicken** [comment]
There are currently two places where data is stored.

1) In VSCode state
2) In a project config file. This file needs to be created and populated manually at this time.

VSCode state can be at the workspace or global. The global setting allows syncing focus groups across devices. 
It is also possible to mix and match these as VSCode has different tiers of settings.

Anyway, the second option is to specify groups in the .filefocus.json file.

From your example try pasting these contents into the file save and refresh the focus groups.

`{
  "store": [
    {
      "name": "Config",
      "path": [".firebaserc", "main.dart"]
    },
    {
      "name": "My Models",
      "path": ["models"]
    }
  ]
}`
